### PR TITLE
[admin] Add release workflow

### PR DESCRIPTION
**Pull request by TylerHelmuth**
Adds a new `release.yaml` workflow that is triggered on the push of new tags.  Shamelessly stolen and modified from https://github.com/keyval-dev/opentelemetry-go-instrumentation/blob/master/.github/workflows/release.yml.

I am woefully underqualified to name this image, but I went with `otel/autoinstrumentation-go` for this PR.  I reviewed the [existing otel images on DockerHub](https://hub.docker.com/u/otel), the [images published by the operator repo](https://github.com/orgs/open-telemetry/packages?repo_name=opentelemetry-operator), and the the java agent .jar name (`opentelemetry-javaagent.jar`).  Someone with authority should name this image because we don't want to rename it in the future.

Tested via my fork.  Published images can be found here:
- https://hub.docker.com/r/thelmuth34/autoinstrumentation-go/tags
- https://github.com/TylerHelmuth/opentelemetry-go-instrumentation/pkgs/container/opentelemetry-go-instrumentation%2Fautoinstrumentation-go

Closes https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/31.

### CODE REVIEW

**Comment by edeNFed** [comment]
Can we start with just amd64 support?
ARM support requires many code changes that need to be merged before we can actually build container images for ARM processors. (already have it in the keyval repo, I will open a separate PR for that soon).

**Comment by TylerHelmuth** [comment]
| Can we start with just amd64 support?

Updated

**Review by MikeGoldsmith — APPROVED** [review]
Looks good to me 👍🏻

**Review by edeNFed — APPROVED** [review]
lgtm

**Comment by TylerHelmuth** [comment]
Added an agenda item to next week's go auto-instrumentation SIG meeting to finalize acceptance from the maintainers on publishing an image and image name.

File: .github/workflows/release.yaml
Comment (pellared) [inline .github/workflows/release.yaml:26]: I suggest avoiding the "Agent" name as it brings confusion. More https://github.com/open-telemetry/opentelemetry.io/issues/1689

can we use `auto` or maybe even `auto-instrumentation` instead?

Comment (pellared) [inline .github/workflows/release.yaml:26]: It may be good to use https://github.com/docker/metadata-action

Example usage: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/.github/workflows/build_and_push_image.yml

Also you can see https://docs.github.com/en/actions/publishing-packages/publishing-docker-images

File: .github/workflows/release.yaml
Comment (pellared) [inline .github/workflows/release.yaml:18]: I think we can simply push it to GitHub registry

**Review by pellared — COMMENTED** [review]
I suggest testing it on a fork before doing any merge.

File: .github/workflows/release.yaml
Comment (TylerHelmuth) [reply .github/workflows/release.yaml:26]: If the community like including `auto`, we could mimic the operator's style and use `otel/autoinstrumentation-go`.

File: .github/workflows/release.yaml
Comment (TylerHelmuth) [reply .github/workflows/release.yaml:18]: I wouldn't excluding pushing to GitHub registry, but there is precedence for using DockerHub for OTel images.  It seems the community is using both locations (operator only used GitHub, demo and collector publish to both).  As long as I have an image to use I'll be happy so I don't care where it lives.  If starting with only Github is easier then that works for me.

Comment (pellared) [reply .github/workflows/release.yaml:18]: I just have no idea if the secrets are available.

Comment (TylerHelmuth) [reply .github/workflows/release.yaml:18]: https://github.com/open-telemetry/community/issues/1377 is completed so the secrets in `DOCKER_USERNAME` and `DOCKER_PASSWORD` should be available to us if we chose to use them.

File: .github/workflows/release.yaml
Comment (TylerHelmuth) [reply .github/workflows/release.yaml:26]: Where is the .Net image being pushed?  I don't see it in GitHub or DockerHub.

Comment (TylerHelmuth) [reply .github/workflows/release.yaml]: Updated name to `autoinstrumentation-go`.

Comment (TylerHelmuth) [reply .github/workflows/release.yaml]: That action is amazing, thanks for pointing it out.

File: .github/workflows/release.yaml
Comment (TylerHelmuth) [reply .github/workflows/release.yaml]: Updated to push to both.

**Comment by TylerHelmuth** [comment]
@pellared thank you for the useful examples.  I have update the action to push to both ghcr.io and DockerHub as directed by https://docs.github.com/en/actions/publishing-packages/publishing-docker-images#publishing-images-to-docker-hub-and-github-packages.  I was able to test this using my fork, images can be found here:
- https://hub.docker.com/r/thelmuth34/autoinstrumentation-go/tags
- https://github.com/TylerHelmuth/opentelemetry-go-instrumentation/pkgs/container/opentelemetry-go-instrumentation%2Fautoinstrumentation-go

@MikeGoldsmith @edeNFed @pellared please re-review.

File: .github/workflows/release.yaml
Comment (pellared) [inline .github/workflows/release.yaml:16]: Can we pin all GitHub Actions' verions?\
E.g. `actions/checkout@v3.3.0`.
This would follow the same pattern as in https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/main/.github/workflows/links.yml

Dependabot is making a PRs to bump and we would mitigate the possible risk if a new version of a GitHub Action would contain a bug.

**Review by pellared — APPROVED** [review]

File: .github/workflows/release.yaml
Comment (MrAlias) [inline .github/workflows/release.yaml:16]:
```suggestion
        uses: actions/checkout@v3.3.0
```

**Pull request merged**
